pgjdbc/src/main/java/org/postgresql/core/ParameterContext.java

```diff
@@ -0,0 +1,202 @@
+
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * {@link org.postgresql.core.Parser} stores information about placeholder occurrences in
+ * ParameterContext. In case of standard JDBC placeholders {@code '?'} the position in the SQL text
+ * is recorded. For named placeholders {@code ":paramName"} the name is recorded as well as the
+ * position. {@code PgPreparedStatement} can then use the name to lookup the parameter corresponding
+ * index. These values are also used by toString() methods to provide a human readable
+ * representation of the SQL text.
+ */
+public class ParameterContext {
+
+  enum BindStyle {
+    POSITIONAL, NAMED
+  }
+
+  /**
+   * EMPTY_CONTEXT is immutable. Calling the add-methods will result in
+   * UnsupportedOperationException begin thrown.
+   */
+  public static final ParameterContext EMPTY_CONTEXT = new ParameterContext() {
+    @Override
+    public int addPositionalParameter(int position) throws SQLException {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int addNamedParameter(int position, String bindName) throws SQLException {
+      throw new UnsupportedOperationException();
+    }
+  };
+
+  private @Nullable BindStyle bindStyle = null;
+  private @Nullable List<Integer> placeholderPositions = null;
+  private @Nullable List<String> placeholderNames = null;
+  private @Nullable List<Integer> placeholderAtPosition = null;
+
+  /**
+   * Adds a positional parameter to this ParameterContext. Once a positional parameter have been
+   * added all subsequent parameters must be positional. Positional parameters cannot be reused, and
+   * their order of appearance will correspond to the parameters sent to the PostgreSQL backend.
+   *
+   * @param position in the SQL text where the parser captured the placeholder.
+   * @return 1-indexed position in the order of appearance of positional parameters
+   * @throws SQLException if positional and named parameters are mixed.
+   */
+  public int addPositionalParameter(@NonNegative int position) throws SQLException {
+    checkAndSetBindStyle(BindStyle.POSITIONAL);
+    int bindIndex = checkAndAddPosition(position);
+
+    if (placeholderAtPosition == null) {
+      placeholderAtPosition = new ArrayList<>();
+    }
+
+    placeholderAtPosition.add(bindIndex);
+
+    return bindIndex + 1;
+  }
+
+  public boolean hasParameters() {
+    return placeholderPositions != null && !placeholderPositions.isEmpty();
+  }
+
+  public boolean hasNamedParameters() {
+    return placeholderNames != null && !placeholderNames.isEmpty();
+  }
+
+  /**
+   * @param i 0-indexed position in the order of first appearance
+   * @return The name of the placeholder at this backend parameter position
+   */
+  public String getPlaceholderName(@NonNegative int i) {
+    if (placeholderNames == null) {
+      throw new IllegalStateException("Call hasNamedParameters() first.");
```

Comment:
Should this be `no placeholders found` rather than an obscure`Call hasNamedParameters() first` ?

Remember, most likely you are talking to application **user** here rather than to the application developer.
In other words:
1) Developer commits bad code and ships the application
2) User launches the app and gets `Call hasNamedParameters() first`. That does not really help since the user can't `call hasNamedParameters`

Reply:
So not exactly sure how either helps the end user? This is still buggy code.

Reply:
This is more or less in line with how we handle result sets that have not called next() 
`ResultSet not positioned properly, perhaps you need to call next.`
is the message from ResultSet